ESPS/general/man/man3/sendwaves.3
.\" Copyright (c) 1991, Entropic Research Lab, Inc.
.\" @(#)sendwaves.3	1.3 06 Jun 1995 ERL
.TH "SEND_XWAVES" 3\-ESPSu 06 Jun 1995
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Speech, Inc.
.SH "NAME"
open_xwaves \- open a socket connection to an \fIxwaves+\fR server
.br
send_xwaves \- send a command to an \fIxwaves+\fR server
.br
close_xwaves \- close a socket connection to an \fIxwaves+\fR server
.br
send_xwaves2 \- send a single command to an \fIxwaves+\fR server without a separate open and close
.SH "SYNOPSIS"
.ft B
#include <esps/esps.h>
#include <esps/ss.h>
.sp
.nf
SOCKET *
open_xwaves(host, port, verbose)
char *host;
int port, verbose;
.sp
int
send_xwaves(sp, str)
SOCKET *sp;
char *str;
.sp
void
close_xwaves(sp)
SOCKET *sp;
.sp
int
send_xwaves2(host, port, str, verbose)
open_xwaves(host, port, verbose)
char *host, *str;
int port, verbose;
.fi
.ft R
.sp
.SH "DESCRIPTION"
.PP
These functions are used to communicate with an \fIxwaves+\fR program
running in server mode.  For more details on this mode of \fIxwaves+\fR
see \fIxwaves+\fR\(\-1\s-1ESPS\s+1).
.PP
The functions \fIopen_xwaves\fR, \fIsend_xwaves\fR, and
\fIclose_xwaves\fR are used in cases where a program is going to send
more than one command to an \fIxwaves+\fR server.   In this case, a
socket connection is maintained open, so that an arbitrary number of
commands can be sent over it.   When the connection is no longer needed,
it is closed.
.PP
The function \fIsend_xwaves2\fR is used in cases where only a single
command is going to be sent.  It combines the open and close function
into a single call.   It should not be used to send a large number of
commands, however, due to the overhead of opening and closing the
connection.
.PP
.I open_xwaves
attempts to open a socket connection to an \fIxwaves+\fR server.  If
\fIhost\fR is non-NULL, then the connection is attempted on the machine
with the hostname of \fIhost\fR.   If \fIhost\fR is NULL and the
environment variable \fBWAVES_HOST\fR is defined, then the connection is
attempted to a server on the hostname specified by \fBWAVES_HOST\fR.
If \fBWAVES_HOST\fR is not defined, then the connection is attempted to
a server on the same has as is running this function (i.e.
\fIlocalhost\fR).   If \fIport\fR is non-zero, then its value is used as
the socket port for the connection attempt.   If \fIport\fR is zero and
the environment variable \fBWAVES_PORT\fR is defined, then the port
specified in \fBWAVES_PORT\fR is used for the attempt.   If
\fBWAVES_PORT\fR is not defined then an ESPS default value (given in
\fIesps.h\fR is used instead).   If \fIverbose\fR is non-zero, then
a message is printed if the connection cannot be opened, mentioning the
hostname and port number.   If \fIverbose\fR is zero, no message is
printed and only a status value is returned.   The return value of the
function is the SOCKET pointer value for success and NULL for failure.
.PP
.I send_xwaves
sends a command to an \fIxwaves+\fR running in server mode.   The value
for \fIsp\fR must be a socket pointer previously returned by
\fIopen_xwaves\fR.   \fIstr\fR is a pointer to the command to send and
must be non-NULL and terminated with a newline character.  
The function returns zero for success and -1 for
failure to communicate.
.PP
.I close_xwaves
closes an existing socket connection to an \fIxwaves+\fR server.   The
value of \fIsp\fR is a socket pointer to a previously opened
\fIxwaves+\fR server.
.PP
.I send_xwaves2
combines the above functions into a single function, to open a server
connection, send a command, and then close a connection.   This is
convienent when only a single command is being sent.   The arguments
\fIhost\fR, \fIport\fR, \fIstr\fR, and \fIverbose\fR have the 
same meaning as with the above functions. \fIstr\fR is a pointer 
to the command to send and must be non-NULL and terminated with a
newline character.  
.SH "EXAMPLE"
.nf
\fI/* Open a connection */\fR
SOCKET *sp;
char buffer[100];

sp = open_xwaves(host, port, 1);
if (sp == NULL)
 \fIerror...\fR

sprintf(buffer,"make file %s\n",filename);
if (send_xwaves(sp, buffer) == -1)
 \fIerror...\fR

close_xwaves(sp);
.fi
.SH DIAGNOSTICS
.PP
An assertion failure occurs if \fIstr\fR or \fIsp\fR is NULL.
.SH "SEE ALSO"
.PP
send_xwaves(1\-\s-1ESPS\s+1), waves_show_sd(3-ESPS)
.SH "AUTHOR"
.PP
Alan Parker, based upon the Simple Socket Library by Mat Watson and
Hubert Bartels.  Used with permission.


ESPS/general/src/lib/sendwaves.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1990-1991  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Alan Parker
 * Checked by:
 * Revised by:
 *
 * Brief description: sends commands to xwaves via a socket connection
 *
 */

static char *sccs_id = "@(#)sendwaves.c	1.3	2/20/96	ERL";


#include <stdio.h>
#include <esps/esps.h>
#if !defined(LINUX_OR_MAC)
#include <esps/unix.h>
#endif
#include <esps/ss.h>

/*
  open_xwaves():
  Trys to open a connection to an xwaves+ server.  First use host and
  port arguments, if they are NULL and/or zero (for port) use environment
  variables WAVES_HOST and/or WAVES_PORT if defined.   If not defined,
  then use defaults from esps/esps.h
*/

extern int debug_level;

SOCKET *
open_xwaves(host, port, verbose)
char *host;
int port,verbose;
{
	int waves_port = DEFAULT_PORT;
        char hostname[MAXHOSTNAMELEN+1];
	SOCKET *sp;

	hostname[0] = '\0';
	if (!port && getenv("WAVES_PORT"))
	  waves_port = atoi(getenv("WAVES_PORT"));
	else if (port > 0)
	  waves_port = port;

	if (!host && getenv("WAVES_HOST"))
	  (void)strcpy(hostname,getenv("WAVES_HOST"));
	else if (host)
	  (void)strcpy(hostname,host);
	else {
	  if(gethostname(hostname,MAXHOSTNAMELEN) == -1) {
	    fprintf(stderr,"open_xwaves: cannot get local hostname.\n");
	    fprintf(stderr,
		"open_xwaves: this indicates a serious network problem.\n");
	    return NULL;
	  }
	}
	
	if (debug_level)
	  fprintf(stderr,"open_xwaves: hostname: %s, port: %d\n",
		hostname,waves_port);

	if ((sp = ConnectSock(hostname, waves_port)) == SS_NULL) {
	  if (verbose) {
		fprintf(stderr,
		 "open_xwaves: Cannot connect to xwaves server at ");
		fprintf(stderr, "host %s, port %d.\n",hostname,waves_port);
	  }
	  return NULL;
	}
	return sp;
}

/* 
   send_xwaves:
   send a message to an open xwaves+ socket, return 0 for OK, -1 for error
*/

int
send_xwaves(sp, str)
SOCKET *sp;
char *str;
{
	spsassert(sp && str,"send_xwaves: sp or str null");
	if (SockWrites(str, sp) == SS_EOF)
	  return -1;
	else
	  return 0;
}

/*
  close_xwaves:
  close an open xwaves+ socket.  
*/
void
close_xwaves(sp)
SOCKET *sp;
{
	spsassert(sp,"close_xwaves: sp is null");
	(void)SockClose(sp);
}

/*
  send_xwaves2:
  open a connection, send str, and then close the connection
*/

int
send_xwaves2(host, port, str, verbose)
char *host, *str;
int port, verbose;
{
	SOCKET *sp;
	int waves_port = DEFAULT_PORT;
        char hostname[MAXHOSTNAMELEN+1];
	spsassert(str,"send_xwaves2: str is null");

	if (!port && getenv("WAVES_PORT"))
	  waves_port = atoi(getenv("WAVES_PORT"));
	else if (port)
	  waves_port = port;

	if (!host && getenv("WAVES_HOST"))
	  (void)strcpy(hostname,getenv("WAVES_HOST"));
	else if (host)
	  (void)strcpy(hostname,host);
	else {
	  if(gethostname(hostname,MAXHOSTNAMELEN) == -1) {
	    fprintf(stderr,"send_xwaves: cannot get local hostname.\n");
	    fprintf(stderr,
		"send_xwaves: this indicates a serious network problem.\n");
	    return -1;
	  }
	}
	
	if (debug_level)
	  fprintf(stderr,"send_xwaves2: hostname: %s, port: %d\n",
		hostname,waves_port);

	if ((sp = ConnectSock(hostname, waves_port)) == SS_NULL) {
	  if (verbose) {
		fprintf(stderr,
		 "open_xwaves: Cannot connect to xwaves server at ");
		fprintf(stderr, "host %s, port %d.\n",hostname,waves_port);
	  }
	  return -1;
	}
	
	if (SockWrites(str, sp) == SS_EOF) return -1;
	(void)SockClose(sp);
}
